dborutils/mongo_client.py

```diff
@@ -211,33 +212,52 @@ def parse_argstring(cls, host_spec, default_host=DIVEBAR_HOST):
         """
         Parses colon-delimited argstring into mongo collection spec
 
-        Returns tuple() or tuple(host, port, database, collection)
+        Returns tuple() or tuple(host, port, database, collection) where host optionally
+        contains username and password.
         """
+        default_port = 27017
+        host = port = database = collection = user_pass = None
 
-        DEFAULT_PORT = 27017
-
-        result = []
-
+        # Parse out host spec string into tuple(host, port, database, collection) result.
         if host_spec:
+            if re.search(r'@', host_spec):
+                user_pass, host_spec = host_spec.split('@')
 
-            result = host_spec.split(':')
+            host_spec_parts = host_spec.split(':')
 
-            if len(result) == 2:
+            if len(host_spec_parts) == 2:
                 # database:collection
                 if not default_host:
+                    raise Exception("Default host is required when provided a database "
+                                    "and collection name only.")
 
-                    raise Exception("Default host is required when provided a database and collection name only.")
+                host = default_host
+                port = default_port
+                database = host_spec_parts[0]
+                collection = host_spec_parts[1]
 
-                result.insert(0, default_host)
-                result.insert(1, DEFAULT_PORT)
-            elif len(result) == 3:
+            elif len(host_spec_parts) == 3:
                 # host:database:collection
-                result.insert(1, DEFAULT_PORT)
-            elif len(result) != 4:
-                # NOT host:port:database:collection
-                result = []
+                port = default_port
+                host = host_spec_parts[0]
+                database = host_spec_parts[1]
+                collection = host_spec_parts[2]
+
+            elif len(host_spec_parts) == 4:
+                # host:port:database:collection
+                host = host_spec_parts[0]
+                port = host_spec_parts[1]
+                database = host_spec_parts[2]
+                collection = host_spec_parts[3]
 
-        return tuple(result)
+            else:
+                raise Exception("Does not match parsable connection string format.")
+
+        # Now that connection string is parsed out, add back the user/pass to the host.
+        if user_pass:
```

Comment:
just a question here. but maybe I missed something following. if user_pass does not exist, seems like it will be None from the split above, do all the len() cases, take care of that?